In [11]:
import numpy as np # mathematical operations and algebra
import pandas as pd # data processing, CSV file I/O
import seaborn as sns # Fancier visualizations
sns.set(color_codes=True)
import matplotlib.pyplot as plt # visualization library
from tree import DesicionTreeNode, buildDecisionTree, buildGraph, evaluateTree
import scipy.stats as stats

IMPORTING DATA

Objective:
-> Import data from CSV file into a padas DataFrame.

In [12]:
data = pd.read_csv (r'diabetes_data_upload.csv')
labels = ['Age','Gender','Polyuria','Polydipsia','sudden weight loss','weakness','Polyphagia',
'Genital thrush','visual blurring','Itching','Irritability','delayed healing','partial paresis',
'muscle stiffness','Alopecia','Obesity','class']
df = pd.DataFrame(data, columns= labels)

DATA CLEANING

Objective:
-> Check if the data contains any null, missing, duplicate.
-> If yes take appopriate action.

In [13]:
print(f'No missing values in data set: {not df.isnull().values.any()}')

No missing values in data set: True


Since we have no missing values, no action is required. Our data is clean and we can proceed with the next step.

In [14]:
# duplicate data

print(f'The total number of dumplicated instances (counting the original) is {sum(df.duplicated(keep=False))}')
print(f'The total number of dumplicated instances  (not counting the original) is: {sum(df.duplicated())}')

The total number of dumplicated instances (counting the original) is 376
The total number of dumplicated instances  (not counting the original) is: 269


At first it seems like 269 out of 520 is a lot of duplicated values.
Lets drop age as it is an obvious tie breaker between two rows. Then recalculate the duplicate values.

In [15]:
df_without_age = df.drop(['Age'], axis=1)
print(f'Number of duplicate instances counting the original is {sum(df_without_age.duplicated(keep=False))}')
print(f'Number of duplicate instances NOT counting the original is {sum(df_without_age.duplicated())}')

Number of duplicate instances counting the original is 407
Number of duplicate instances NOT counting the original is 305


It's a lotmore now (as expected) so let's take a look at some of these duplicate values:

In [16]:
duplicated_data = df[df.duplicated()].sort_values(by='Age')
print(duplicated_data[1:5])

     Age Gender Polyuria Polydipsia sudden weight loss weakness Polyphagia  \
374   27   Male       No         No                 No       No         No   
286   27   Male       No         No                 No       No         No   
465   27   Male       No         No                 No       No         No   
474   27   Male       No         No                 No       No         No   

    Genital thrush visual blurring Itching Irritability delayed healing  \
374             No              No      No           No              No   
286             No              No      No           No              No   
465             No              No      No           No              No   
474             No              No      No           No              No   

    partial paresis muscle stiffness Alopecia Obesity     class  
374              No               No       No      No  Negative  
286              No               No       No      No  Negative  
465              No               N

So most of these instances are simply copies of a few 'common' cases. This can be illustrated if we take a look at some of these instances.

same something about why we used chi-2 etc.

In [17]:
df_without_age = df.drop(['Age'], axis=1)
chi2 = 10.828 #chi2 at df 1 ( for 2x2 table) 0.999 confidence
chi2_matrix = []
for label_i in df_without_age.keys():
    chi2_row = []
    for label_j in df_without_age.keys():
        # first generate a contingency table
        contingency_tab = pd.crosstab(df_without_age[label_i], df_without_age[label_j])
        # calculate the chi square value
        chi, p, dof, expected = stats.chi2_contingency(contingency_tab)
        chi2_row.append(chi>chi2)
    chi2_matrix.append(chi2_row)
labels_without_age = labels[1:]
contingency_df = pd.DataFrame(data=chi2_matrix,index=labels_without_age, columns=labels_without_age)
print(contingency_df)

                    Gender  Polyuria  Polydipsia  sudden weight loss  \
Gender                True      True        True                True   
Polyuria              True      True        True                True   
Polydipsia            True      True        True                True   
sudden weight loss    True      True        True                True   
weakness             False      True        True                True   
Polyphagia            True      True        True                True   
Genital thrush        True     False       False               False   
visual blurring       True      True        True               False   
Itching              False     False       False               False   
Irritability         False      True        True               False   
delayed healing      False      True       False               False   
partial paresis       True      True        True                True   
muscle stiffness     False      True        True               F

DECISION TREE METHOD

In [18]:
# discretize the age into a few categories
df_discretize = df.copy(deep=True)
minAge = df_discretize.Age.min()
maxAge = df_discretize.Age.max()
range = maxAge - minAge
df_discretize.Age = pd.cut(df["Age"],
       bins=[minAge, minAge + range/3, minAge + 2*range/3, maxAge], 
       labels=["Young", "Adult", "Old"])

In [19]:
# divide the discretized data into training and test 
df_test = df_discretize.sample(n = 50, replace = False)
df_training = df_discretize.copy(deep=True)
df_training = df_training.drop(df_test.index)

In [20]:
# decision Tree
root = buildDecisionTree(data=df_training, classAttribute='class')
# draw it as a graph
buildGraph(root).view()
# calculate its accuracy
print("Accuracy: ",evaluateTree(root,df_test,'class'))

Accuracy:  0.96


KNN METHOD

In [21]:
df_training


Age  Gender Polyuria Polydipsia sudden weight loss weakness Polyphagia  \
0    Young    Male       No        Yes                 No      Yes         No   
1    Adult    Male       No         No                 No      Yes         No   
2    Adult    Male      Yes         No                 No      Yes        Yes   
3    Adult    Male       No         No                Yes      Yes        Yes   
4    Adult    Male      Yes        Yes                Yes      Yes        Yes   
..     ...     ...      ...        ...                ...      ...        ...   
514  Adult  Female      Yes        Yes                Yes      Yes        Yes   
515  Young  Female      Yes        Yes                Yes       No        Yes   
516  Adult  Female      Yes        Yes                Yes      Yes        Yes   
518  Young  Female       No         No                 No      Yes         No   
519  Adult    Male       No         No                 No       No         No   

    Genital thrush visual blurring Itching Irritability delayed healing  \
0               No              No     Yes           No             Yes   
1               No             Yes      No           No              No   
2               No              No     Yes           No             Yes   
3              Yes              No     Yes           No             Yes   
4               No             Yes     Yes          Yes             Yes   
..             ...             ...     ...          ...             ...   
514             No              No      No           No              No   
515             No              No     Yes           No             Yes   
516             No              No     Yes          Yes             Yes   
518             No             Yes     Yes           No             Yes   
519             No              No      No           No              No   

    partial paresis muscle stiffness Alopecia Obesity     class  
0                No              Yes      Yes     Yes  Positive  
1               Yes               No      Yes      No  Positive  
2                No              Yes      Yes      No  Positive  
3                No               No       No      No  Positive  
4               Yes              Yes      Yes     Yes  Positive  
..              ...              ...      ...     ...       ...  
514             Yes               No       No      No  Positive  
515             Yes               No       No      No  Positive  
516             Yes               No       No      No  Positive  
518              No               No      Yes      No  Negative  
519              No               No       No      No  Negative  

[470 rows x 17 columns]